### PR TITLE
Move code into datasets.jl, readwrite.jl, and types.jl

### PR DESCRIPTION
**Pull request by mkitti**
This is my attempt to move dataset code into datasets.jl for #945 

I ended up needing three new files:

* types.jl - high level struct and union definitions that need to be defined early
* datasets.jl - dataset specific code
* readwrite.jl - read and write code common to datasets and attributes

### CODE REVIEW

File: src/types.jl [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+Base.length(::Type{FixedString{N,PAD}}) where {N,PAD} = N
+Base.length(str::FixedString) = length(typeof(str))
+pad(::Type{FixedString{N,PAD}}) where {N,PAD} = PAD
+pad(x::T) where {T<:FixedString} = pad(T)
+
+struct VariableArray{T}
+    len::Csize_t
+    p::Ptr{Cvoid}
+end
+Base.eltype(::Type{VariableArray{T}}) where T = T
```
Comment (simonbyrne) [inline src/types.jl:186]: I think this part should be moved to a different file: these aren't types corresponding to HDF5 objects, but are instead convenience wrappers matching certain types of data (in fact, I don't think some of them should exist, but that's another story)

Comment (simonbyrne) [reply src/types.jl:186]: perhaps move this back to `typeconversions.jl`?

**Review by simonbyrne — APPROVED** [review]
Overall I think this structure makes the most sense. Defining HDF5 types up front will make some organization pieces easier.

File: src/types.jl (hunk 1 above)
Comment (mkitti) [reply src/types.jl]: Moved back to typeconversions.jl in 565fcaa

**Comment by musm** [comment]
LGTM will merge later today sans objections

**Comment by mkitti** [comment]
Looks good from my side as well, so merging. Let's move on to the next things.

**Pull request merged**